x/cosmwasmpool/README.md

```diff
@@ -0,0 +1,284 @@
+
+Common interface and later be defined for the contract to implement as spec and/or create a separated crate for that purpose.
+
+It's important to note that the _**contract itselfs hold tokens that are provided by users**_.
+
+
+## Swap
+
+One of the main reason why CosmWasm pool is implemented as a module + contract rather than a contract only is that it allows us to use the existing pool manager module to handle swap, which means things like swap routing, cross chain swap, and other functionality that depends on existing pool interface works out of the box.
+
+```mermaid
+graph TD;
+  Sender((Sender))
+  Sender -- swap --> x/poolmanager
+  x/poolmanager -- route msg to --> x/cosmwasmpool
+  x/cosmwasmpool -- sudo execute contract --> x/wasm
+  x/wasm -- sudo --> wasm/pool
+
+  x/cosmwasmpool -- send token_in from sender to wasm/pool --> x/bank
+  wasm/pool -- send token_out to sender --> x/bank
+```
+
+Pool contract's sudo endpoint expect the following message variant:
+
+```rs
+/// SwapExactAmountIn swaps an exact amount of tokens in for as many tokens out as possible.
+/// The amount of tokens out is determined by the current exchange rate and the swap fee.
+/// The user specifies a minimum amount of tokens out, and the transaction will revert if that amount of tokens
+/// is not received.
+SwapExactAmountIn {
+    sender: String,
+    token_in: Coin,
+    token_out_denom: String,
+    token_out_min_amount: Uint128,
+    swap_fee: Decimal,
+},
+/// SwapExactAmountOut swaps as many tokens in as possible for an exact amount of tokens out.
+/// The amount of tokens in is determined by the current exchange rate and the swap fee.
+/// The user specifies a maximum amount of tokens in, and the transaction will revert if that amount of tokens
+/// is exceeded.
+SwapExactAmountOut {
+    sender: String,
+    token_in_denom: String,
+    token_in_max_amount: Uint128,
+    token_out: Coin,
+    swap_fee: Decimal,
+},
+```
+
+The reason why this needs to be sudo endpoint, which can only be called by the chain itself, is that the chain can provide correct information about `swap_fee`, which can be deviated from contract defined `swap_fee` in multihop scenario.
+
+(TBD)
+Problem with this approach is that swap expects funds to be sent with the same amount as `token_in` but there is no way to validate that in contract, so we need to make sure of that in `x/cosmwasmpool` module.
+
+Alternative approach is to expose swap in execute endpoint instead and validate that msg sender is a specific module account so that we can leverage `MsgExecuteContract`'s `funds` field to send funds to the contract.
```

Comment:
CC: @nicolaslara

Reply:
I don't understand the benefit of the `MsgExecture` approach. The sudo endpoint seems to be better from a security standpoint. 

The only benefit of having it as execute is that users would be able to interact with the contract directly. However, that is blocked anyways because we provide swap fee as param and must validate that sender is a module account

Reply:
I wasn't aware of the swap fee issue. I thought that could be calculated or queried from the contract (but now I see why it can't).

The benefit was that the contract would be standalone and the module would be just a wrapper that doesn't need to do custom logic (send the tokens to the contract before calling execute); we could then let anyone execute the contract directly. But with the swap-fee issue in mind, then I agree that the sudo solution is better

Reply:
Seems that everyone is on the same page now, will remove TBD section then.

Reply:
Sweet